data/blog/improving-nodejs-loader-performance.md

```diff
@@ -0,0 +1,180 @@
+- `pkg.name` is used if there is a self referencing require/import.
+
+Additionally, Node.js supports an experimental version of `Subresource Integrity` checkwhich uses the result of this package.json to validate the integrity of the file.
+
+The most important usage is that, for every `require/import` call, Node.js needs to know the module format of the file. For example, if the user require's a NPM module that uses ESM on a CommonJS (CJS) application, Node.js will need to parse the `package.json` file of that module and throw an error if the NPM package is ESM.
+
+Because of all of these calls and usages across ESM and CJS loaders, `package.json` reader is one of the most important parts of the Node.js loader implementation.
+
+## Optimizations
+
+### Optimizing caching layer
+
+In order to optimize the `package.json` reader performance, I first moved the caching layer to the C++ side to make the implementation be closer to the filesystem call as much as possible. This decision forced to parse the JSON file in C++. At this point, I had 2 options:
+
+- Use V8's `v8::JSON::Parse()` method which takes a `v8::String` as an input and returns a `v8::Value` as an output.
+- Use `simdjson` library to parse the JSON file.
+
+Since the filesystem returns a string, converting that string into a `v8::String` just to retrieve the keys and values as a `std::string` didn't make sense. Therefore, I added `simdjson` as a dependency to Node.js and used it to parse the JSON file. This change enabled us to parse the JSON file in C++ and extract and return only the necessary fields to the JavaScript side, reducing the size of the input that needs to be serialized/deserialized.
+
+### Avoiding serialization cost
+
+In order to avoid returning unnecessary large objects, I changed the signature of the `readPackage` function to return only the necessary fields. This change simplified the `shouldUseESMLoader` as follows:
+
+```js
+function shouldUseESMLoader(mainPath) {
+  // Determine the module format of the entry point.
+  if (mainPath && mainPath.endsWith('.mjs')) { return true; }
+  if (!mainPath || mainPath.endsWith('.cjs')) { return false; }
+
+  const response = getNearestParentPackageJSONType(mainPath);
+
+  // No package.json or no `type` field.
+  if (response === undefined || response[0] === 'none') {
+    return false;
+  }
+
+  const {
+    0: type,
+    1: filePath,
+    2: rawContent,
+  } = response;
+
+  checkPackageJSONIntegrity(filePath, rawContent);
+
+  return type === 'module';
+}
+```
+
+Moving the caching layer to C++ enabled us to expose micro-functions that returns enums (integers) instead of strings to get a type of a `package.json` file.
+
+### Reducing C++ calls to 1 to 1
+
+The same function that was mentioned above but on ESM loader called `getPackageScopeConfig` made a lot of C++ calls in order to resolve and retrieve the applicable `package.json` file. The implementation was as follows:
```

Comment:
rearranging this sentence might help. Where is it mentioned above?
```suggestion
The same `getPackageScopeConfig` function mentioned above doing ESM loading made a lot of C++ calls in order to resolve and retrieve the applicable `package.json` file. The implementation was as follows:
```

Reply:
On CJS it is `readPackageConfig` but on ESM that function name is `getPackageScopeConfig`. That's what I wanted to mention here. The recommended comment is wrong.